### ISSUE TITLE
Some devices won't run tests: "java.lang.VerifyError: Rejecting class X because it failed compile-time verification"

### ISSUE THREAD

**Issue by gefilte** [tmurakami/dexopener#3]
This is a strange one. I am running the same test package including dexopener on different devices. For some the tests run normally and for others the tests won't start because of this error. Here is the entire output for the test run on one of the failing devices:

```
06-01 14:46:57.454 I/art: Late-enabling -Xcheck:jni
06-01 14:46:57.454 I/art: VMHOOK: rlim_cur : 0 pid:4547
06-01 14:46:57.489 I/art: Debugger is no longer active
06-01 14:46:57.589 I/MultiDex: VM with version 2.1.0 has multidex support
06-01 14:46:57.589 I/MultiDex: install
06-01 14:46:57.590 I/MultiDex: VM has multidex support, MultiDex support library is disabled.
06-01 14:46:58.992 I/art: Background sticky concurrent mark sweep GC freed 177210(11MB) AllocSpace objects, 19(708KB) LOS objects, 19% free, 41MB/51MB, paused 5.139ms total 72.868ms
06-01 14:47:00.495 I/art: Background sticky concurrent mark sweep GC freed 216115(13MB) AllocSpace objects, 19(708KB) LOS objects, 22% free, 42MB/54MB, paused 6.577ms total 76.864ms
06-01 14:47:01.277 I/art: override thread count:-j4
06-01 14:47:02.687 I/art: override thread count:-j4
06-01 14:47:02.988 W/art: Verification of   .network.ClientCallback com.redacted.activity.ActivityTestBase.getClientCallback() took 1.321s
06-01 14:47:04.332 I/art: override thread count:-j4
06-01 14:47:04.653 W/art: Verification of void com.redacted.activity.ActivityTestBase.setUp() took 1.661s
06-01 14:47:04.654 E/AndroidRuntime: FATAL EXCEPTION: main
                                     Process: com.redacted.test, PID: 4547
                                     java.lang.VerifyError: Rejecting class com.redacted.activity.RedactedActivityTest because it failed compile-time verification (declaration of 'com.redacted.activity.RedactedActivityTest' appears in /data/data/com.redacted.test/code_cache/dexopener/classes-1967783184.zip)
                                         at java.lang.reflect.Constructor.newInstance(Native Method)
                                         at junit.framework.TestSuite.createTest(TestSuite.java:61)
                                         at junit.framework.TestSuite.addTestMethod(TestSuite.java:294)
                                         at junit.framework.TestSuite.addTestsFromTestCase(TestSuite.java:150)
                                         at junit.framework.TestSuite.<init>(TestSuite.java:129)
                                         at junit.runner.BaseTestRunner.getTest(BaseTestRunner.java:118)
                                         at android.test.AndroidTestRunner.getTest(AndroidTestRunner.java:149)
                                         at android.test.AndroidTestRunner.setTestClassName(AndroidTestRunner.java:57)
                                         at android.test.suitebuilder.TestSuiteBuilder.addTestClassByName(TestSuiteBuilder.java:78)
                                         at android.test.InstrumentationTestRunner.parseTestClass(InstrumentationTestRunner.java:444)
                                         at android.test.InstrumentationTestRunner.parseTestClasses(InstrumentationTestRunner.java:425)
                                         at android.test.InstrumentationTestRunner.onCreate(InstrumentationTestRunner.java:371)
                                         at com.android.test.runner.MultiDexTestRunner.onCreate(MultiDexTestRunner.java:31)
                                         at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4991)
                                         at android.app.ActivityThread.access$1600(ActivityThread.java:150)
                                         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1455)
                                         at android.os.Handler.dispatchMessage(Handler.java:102)
                                         at android.os.Looper.loop(Looper.java:168)
                                         at android.app.ActivityThread.main(ActivityThread.java:5781)
                                         at java.lang.reflect.Method.invoke(Native Method)
                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:797)
                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:687)

```

We have successes on a Samsung Galaxy S6 Marshmallow, a Google Nexus 6 Lollipop, and a Google Pixel C Nougat.
We have failures on a Samsung Galaxy S7 Marshmallow and an HTC One A9 Marshmallow.
So there doesn't seem to be a pattern of why one device should pass and the other fail.

**Comment by tmurakami** [comment]
Are you testing an Android library?
Or an app?
If you are testing an Android app, it seems strange that the test class is in DexOpener's dex cache.

Could you add the following three lines to your `AndroidJUnitRunner#newApplication` and show me the logs?
```java
Log.d("YourApp", "context.getApplicationInfo().sourceDir=" + context.getApplicationInfo().sourceDir);
Log.d("YourApp", "getContext().getApplicationInfo().sourceDir=" + getContext().getApplicationInfo().sourceDir);
Log.d("YourApp", "getTargetContext().getApplicationInfo().sourceDir=" + getTargetContext().getApplicationInfo().sourceDir);
```

**Comment by gefilte** [comment]
The module under test is a library (as in `com.android.library` gradle plugin) but it's just a module attached to an app project. I was able to reproduce in two different modules with the same result.

Here's the output from your requested debugging:
```
06-02 11:11:21.034 D/TestRunner: context.getApplicationInfo().sourceDir=/data/app/com.redacted.test-2/base.apk
06-02 11:11:21.034 D/TestRunner: getContext().getApplicationInfo().sourceDir=/data/app/com.redacted.test-2/base.apk
06-02 11:11:21.034 D/TestRunner: getTargetContext().getApplicationInfo().sourceDir=/data/app/com.redacted.test-2/base.apk
```

**Comment by tmurakami** [comment]
Thank you for your reply.

There is one more request.
Could I have all the logs?
If it is impossible, would you please show me the logs tagged as `dex2oat`?

**Comment by gefilte** [comment]
I created a gist:
https://gist.github.com/gefilte/7f4d0d6a487c085896dc86694e0908d1

It's a lot of noise but should include what you are looking for. It is unfiltered logcat starting at the point I install the app package through the completed run of the test. All the relevant package references will contain `com.redacted`, and I did find/replace so that the relevant modules show up separately (packageA, packageB, appPackage, authlib, and anotherpackage. Great names, eh?) The test is in packageA, but depends on some things in packageB. The overall packageA depends on packageB, authlib, and anotherpackage.

Anyway, hopefully you find what you're looking for.

**Comment by tmurakami** [comment]
Thank you.
I checked your log file and it seems to be a problem of ASMDEX which is a bytecode manipulation library used by DexOpener.

```
06-02 16:26:55.675 I/dex2oat: Verification error in void com.redacted.packageA.util.MockPackageBManagerModule.<init>()
06-02 16:26:55.675 I/dex2oat: void com.redacted.packageA.util.MockPackageBManagerModule.<init>(): [0xFFFFFFFF] invalid reg index 1+25 in range invoke (> 25)
06-02 16:26:55.675 W/dex2oat: Verification failed on class com.redacted.packageA.util.MockPackageBManagerModule in /data/data/com.redacted.packageA.test/code_cache/dexopener/classes114178148.zip because: Verifier rejected class com.redacted.packageA.util.MockPackageBManagerModule due to bad method void com.redacted.packageA.util.MockPackageBManagerModule.<init>()
```

The `dex2oat` which is a DEX code compilation tool failed bytecode verification.
According to the second line, register size overflow has occurred.

ASMDEX is an old library, so we plan to change it to other newer libraries, but it will take time.
If you cannot wait, we recommend [kotlin-allopen](https://kotlinlang.org/docs/reference/compiler-plugins.html#all-open-compiler-plugin) plugin instead.

**Comment by tmurakami** [comment]
We have released a new version 0.9.9 using Smali instead of ASMDEX.
This problem may be solved, so please try it.

**Comment by gefilte** [comment]
This seems to have solved the problem. Thanks!
